### PR TITLE
Implement -P for parameter overrides

### PR DESCRIPTION
**Pull request by kba**
Part of #514 now.

### CODE REVIEW

**Comment by tboenig** [comment]
It looks like a good solution.

File: ocrd/bashlib/src/parse_argv.bash [hunk 1]
```diff
@@ -19,13 +19,15 @@ ocrd__parse_argv () {
...
     ocrd__argv[overwrite]=false
+    ocrd__argv[parameter_override]=""
 
     while [[ "${1:-}" = -* ]];do
         case "$1" in
             -l|--log-level) ocrd__argv[log_level]=$2 ; shift ;;
             -h|--help|--usage) ocrd__usage; exit ;;
             -J|--dump-json) ocrd__dumpjson; exit ;;
             -p|--parameter) ocrd__argv[parameter]="$2" ; shift ;;
+            -P|--parameter-override) ocrd__argv[parameter_override]+=" -P $2 $3" ; shift ; shift ;;
```
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/bashlib/src/parse_argv.bash:30]: Unfortunately bash does not allow using an array inside an array variable:
```
bash: ocrd__argv[parameter_override]: cannot assign list to array member
```

So I believe the only way to get quotation right in bashlib would be to introduce additional quotes...

```suggestion
            -P|--parameter-override) ocrd__argv[parameter_override]+="-P '$2' '$3'"; shift ; shift ;;
```

...which then get "eaten up" by the _unquoted_ expansion you already used in your `parse-params` call. Have not tested it, though.

[the remark this answers is not in the document]
Comment (kba) [reply ocrd/bashlib/src/parse_argv.bash:30]: i tried that first but the quotes were not handled properly. I'll try again later.

[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/bashlib/src/parse_argv.bash:30]: I see. Too bad! 

Notice that we _also_ need a solution for multi-valued option passing in `-p|--parameter` now. This currently just overwrites the previous value. So we should also concatenate a long strong with `-p $2` each time.

Perhaps `eval` can help, or using `@Q` expansion as in `... parse-params ${ocrd__argv[parameter]@Q} ${ocrd__argv[parameter_override]@Q}`?

[the remark this answers is not in the document]
Comment (kba) [reply ocrd/bashlib/src/parse_argv.bash:30]: ...or just define local arrays for both parameters and overrides. repeatable params are another PR iirc.

[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/bashlib/src/parse_argv.bash:30]: > ...or just define local arrays for both parameters and overrides.

Yes, `ocrd__argv` is probably only going to be used internally anyway. With separate arrays you get a clean solution.

[the remark this answers is not in the document]
Comment (kba) [reply ocrd/bashlib/src/parse_argv.bash]: I solve it with two local arrays, so now you can do

ocrd-im6convert -I FOO -O BAR -P output-format "image/png this addition makes the value invalid!"

File: ocrd/ocrd/decorators.py [hunk 2]
```diff
@@ -27,6 +28,13 @@ def _set_root_logger_version(ctx, param, value):    # pylint: disable=unused-arg
                                 default='{}',
                                 callback=lambda ctx, param, value: parse_json_string_or_file(value))
```
Comment (bertsky) [inline ocrd/ocrd/decorators.py:29]:
```suggestion
                                multiple=True,
                                callback=lambda ctx, param, value: parse_json_string_or_file(value))
```

Isn't that still missing?

File: ocrd/ocrd/cli/ocrd_tool.py [hunk 3]
```diff
@@ -119,12 +121,14 @@ def ocrd_tool_tool_dump(ctx):
...
 @parameter_option
+@parameter_override_option
 @click.option('-j', '--json', help='Output JSON instead of shell variables', is_flag=True, default=False)
 @pass_ocrd_tool
-def ocrd_tool_tool_parse_params(ctx, parameter, json):
+def ocrd_tool_tool_parse_params(ctx, parameter, parameter_override, json):
     """
     Parse parameters with fallback to defaults and output as shell-eval'able assignments to params var.
     """
+    set_json_key_value_overrides(parameter, *parameter_override)
```
Comment (bertsky) [inline ocrd/ocrd/cli/ocrd_tool.py:131]: I guess here also something must still be done to adapt for multiple `-p` (list-valued `parameter`), right?

File: ocrd/bashlib/src/parse_argv.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/bashlib/src/parse_argv.bash]: Splendid!

File: ocrd/ocrd/decorators.py (hunk 2 above)
Comment (kba) [reply ocrd/ocrd/decorators.py:29]: That's in the other params PR https://github.com/OCR-D/core/pull/514/files

File: ocrd/ocrd/cli/ocrd_tool.py (hunk 3 above)
Comment (kba) [reply ocrd/ocrd/cli/ocrd_tool.py:131]: Yes, I'll merge this into https://github.com/OCR-D/core/pull/514 now to get both repeatable parameters and parameter overrides in v2.11.0

Comment (kba) [reply ocrd/ocrd/cli/ocrd_tool.py:131]: Actually, that is handled in the click callback already, `kwargs["parameter"]` is already merged (in #514, that is).

Comment (bertsky) [reply ocrd/ocrd/cli/ocrd_tool.py:131]: Oh, right, already forgot about that!

**Comment by mikegerber** [comment]
Also thanks for this, this makes working with the CLI less clunky :+1:

**Pull request merged**